File: src/services/apiClient.ts
/**
 * API Client Service for FastAPI Backend
 * Provides typed HTTP client methods for all VRI Digital Twin API endpoints
 */

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

interface ApiResponse<T> {
  data?: T;
  error?: string;
  status: number;
}

class APIClient {
  private baseURL: string;

  constructor(baseURL: string = API_URL) {
    this.baseURL = baseURL;
  }

  private async request<T>(
    method: string,
    endpoint: string,
    body?: any
  ): Promise<ApiResponse<T>> {
    try {
      const url = `${this.baseURL}${endpoint}`;
      const options: RequestInit = {
        method,
        headers: {
          'Content-Type': 'application/json',
        },
      };

      if (body) {
        options.body = JSON.stringify(body);
      }

      const response = await fetch(url, options);

      if (!response.ok) {
        const error = await response.text();
        return {
          error: `HTTP ${response.status}: ${error}`,
          status: response.status,
        };
      }

      const data = await response.json();
      return {
        data,
        status: response.status,
      };
    } catch (error) {
      return {
        error: error instanceof Error ? error.message : 'Unknown error',
        status: 0,
      };
    }
  }

  // Health & Status
  async getHealth() {
    return this.request('/health', 'GET');
  }

  async getReadiness() {
    return this.request('/ready', 'GET');
  }

  // Fields Endpoints
  async getFields() {
    return this.request('/api/v1/fields', 'GET');
  }

  async getField(fieldId: string) {
    return this.request(`/api/v1/fields/${fieldId}`, 'GET');
  }

  async getFieldZones(fieldId: string) {
    return this.request(`/api/v1/fields/${fieldId}/zones`, 'GET');
  }

  async getZone(fieldId: string, zoneId: string) {
    return this.request(`/api/v1/fields/${fieldId}/zones/${zoneId}`, 'GET');
  }

  // Sensors Endpoints
  async getSensors() {
    return this.request('/api/v1/sensors', 'GET');
  }

  async getSensor(sensorId: string) {
    return this.request(`/api/v1/sensors/${sensorId}`, 'GET');
  }

  async getZoneSensors(zoneId: string) {
    return this.request(`/api/v1/sensors/zone/${zoneId}`, 'GET');
  }

  async ingestTelemetry(data: any) {
    return this.request('/api/v1/sensors/telemetry/ingest', 'POST', data);
  }

  // RL Engine Endpoints
  async inferVRIRates(observation: any) {
    return this.request('/api/v1/rl-engine/infer-vri-rates', 'POST', observation);
  }

  async getDecisions(fieldId?: string, status?: string) {
    let endpoint = '/api/v1/rl-engine/decisions';
    const params = new URLSearchParams();
    if (fieldId) params.append('field_id', fieldId);
    if (status) params.append('status_filter', status);
    if (params.toString()) endpoint += `?${params.toString()}`;
    return this.request(endpoint, 'GET');
  }

  async approveDecision(decisionId: string, approvedBy: string) {
    return this.request(
      `/api/v1/rl-engine/decisions/${decisionId}/approve`,
      'POST',
      { approvedBy }
    );
  }

  async overrideDecision(decisionId: string, overrideData: any) {
    return this.request(
      `/api/v1/rl-engine/decisions/${decisionId}/override`,
      'POST',
      overrideData
    );
  }

  async retrainPolicy(trainingData: any) {
    return this.request('/api/v1/rl-engine/retrain-policy', 'POST', trainingData);
  }

  // Irrigation Control Endpoints
  async executeDecision(decision: any) {
    return this.request('/api/v1/irrigation/execute-decision', 'POST', decision);
  }

  async manualTrigger(triggerData: any) {
    return this.request('/api/v1/irrigation/manual-trigger', 'POST', triggerData);
  }

  async getIrrigationStatus(fieldId?: string) {
    let endpoint = '/api/v1/irrigation/status';
    if (fieldId) endpoint += `?field_id=${fieldId}`;
    return this.request(endpoint, 'GET');
  }

  async emergencyStop(fieldId: string) {
    return this.request('/api/v1/irrigation/emergency-stop', 'POST', { fieldId });
  }

  async getIrrigationSchedule(fieldId?: string) {
    let endpoint = '/api/v1/irrigation/schedule';
    if (fieldId) endpoint += `?field_id=${fieldId}`;
    return this.request(endpoint, 'GET');
  }

  // Reports Endpoints
  async generateReport(reportConfig: any) {
    return this.request('/api/v1/reports/generate', 'POST', reportConfig);
  }

  async getReportStatus(reportId: string) {
    return this.request(`/api/v1/reports/${reportId}/status`, 'GET');
  }

  async downloadReport(reportId: string) {
    return this.request(`/api/v1/reports/${reportId}/download`, 'GET');
  }

  async getReportHistory(fieldId?: string, limit: number = 20) {
    let endpoint = '/api/v1/reports/history';
    const params = new URLSearchParams();
    if (fieldId) params.append('field_id', fieldId);
    params.append('limit', limit.toString());
    if (params.toString()) endpoint += `?${params.toString()}`;
    return this.request(endpoint, 'GET');
  }

  async scheduleReport(scheduleConfig: any) {
    return this.request('/api/v1/reports/schedule', 'POST', scheduleConfig);
  }

  async exportWhatIfScenario(scenarioData: any) {
    return this.request('/api/v1/reports/what-if-export', 'POST', scenarioData);
  }
}

// Export singleton instance
export const apiClient = new APIClient();

// Export class for testing
export { APIClient };

File: frontend_streamlit/api_client.py
"""
Cliente HTTP para la API FastAPI.
Mapea IDs locales (zone-nw) a IDs del backend (zone-1-nw).
Expone last_error / is_online para el banner de modo local.
"""
import os
import requests
from dotenv import load_dotenv

load_dotenv()

API_BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8000")
TIMEOUT = float(os.getenv("API_TIMEOUT", "4"))

ZONE_TO_API = {
    "zone-nw": "zone-1-nw",
    "zone-ne": "zone-2-ne",
    "zone-sw": "zone-3-sw",
    "zone-se": "zone-4-se",
}

_last_error: str | None = None
_online: bool | None = None


def last_error() -> str | None:
    return _last_error


def is_online() -> bool:
    global _online
    if _online is None:
        get_health()
    return bool(_online)


def to_api_zone(zone_id: str) -> str:
    return ZONE_TO_API.get(zone_id, zone_id)


def _fail(exc: Exception):
    global _last_error, _online
    _last_error = str(exc)
    _online = False
    return None


def _ok(payload):
    global _last_error, _online
    _last_error = None
    _online = True
    return payload


def _get(endpoint: str, params: dict = None) -> dict | None:
    try:
        response = requests.get(f"{API_BASE_URL}{endpoint}", params=params, timeout=TIMEOUT)
        response.raise_for_status()
        return _ok(response.json())
    except Exception as e:
        return _fail(e)


def _post(endpoint: str, payload: dict) -> dict | None:
    try:
        response = requests.post(f"{API_BASE_URL}{endpoint}", json=payload, timeout=TIMEOUT)
        response.raise_for_status()
        return _ok(response.json())
    except Exception as e:
        return _fail(e)


def get_health() -> dict | None:
    return _get("/health")


def get_fields() -> dict | None:
    return _get("/api/v1/fields/")


def get_field(field_id: str) -> dict | None:
    return _get(f"/api/v1/fields/{field_id}")


def get_field_zones(field_id: str) -> dict | None:
    return _get(f"/api/v1/fields/{field_id}/zones")


def get_sensors() -> dict | None:
    return _get("/api/v1/sensors/")


def get_zone_sensors(zone_id: str) -> dict | None:
    return _get(f"/api/v1/sensors/zone/{to_api_zone(zone_id)}")


def ingest_telemetry(readings: list) -> dict | None:
    return _post("/api/v1/sensors/telemetry/ingest", {"readings": readings})


def infer_vri_rates(zone_id: str, moisture: float, canopy_temp: float) -> dict | None:
    payload = {
        "zoneId": to_api_zone(zone_id),
        "soilMoisture10cm": moisture,
        "canopyTemperature": canopy_temp,
    }
    return _post("/api/v1/rl-engine/infer-vri-rates", payload)


def approve_decision(decision_id: str, approved_by: str) -> dict | None:
    return _post(
        f"/api/v1/rl-engine/decisions/{decision_id}/approve",
        {"approved_by": approved_by},
    )


def override_decision(decision_id: str, depth_mm: float, reason: str) -> dict | None:
    return _post(
        f"/api/v1/rl-engine/decisions/{decision_id}/override",
        {"depthMm": depth_mm, "reason": reason},
    )


def get_irrigation_status(field_id: str = "field-001") -> dict | None:
    return _get("/api/v1/irrigation/status", params={"field_id": field_id})


def get_irrigation_schedule(field_id: str = "field-001") -> dict | None:
    return _get("/api/v1/irrigation/schedule", params={"field_id": field_id})


def execute_irrigation(decision_id: str, zone_id: str, depth_mm: float) -> dict | None:
    payload = {
        "decisionId": decision_id,
        "zoneId": to_api_zone(zone_id) if zone_id else zone_id,
        "depthMm": depth_mm,
    }
    return _post("/api/v1/irrigation/execute-decision", payload)


def manual_trigger(zone_id: str, depth_mm: float, triggered_by: str = "Operador") -> dict | None:
    payload = {
        "zoneId": to_api_zone(zone_id),
        "depthMm": depth_mm,
        "triggeredBy": triggered_by,
    }
    return _post("/api/v1/irrigation/manual-trigger", payload)


def emergency_stop(field_id: str = "field-001") -> dict | None:
    return _post("/api/v1/irrigation/emergency-stop", {"field_id": field_id})


def generate_report(title: str, fmt: str, field_id: str) -> dict | None:
    payload = {"title": title, "format": fmt, "fieldId": field_id}
    return _post("/api/v1/reports/generate", payload)


def get_report_history(field_id: str = "field-001") -> dict | None:
    return _get("/api/v1/reports/history", params={"field_id": field_id})
